test: UserRepository 테스트 코드 작성 #73

wisdom08 opened the pull request:
## 한 일
- #73 이슈 작성 후 UserRepository 테스트 코드를 작성함

## 할 일
- 다른 리포지토리 테스트 작성
- 서비스 테스트코드 작성
- 컨트롤러 테스트코드 작성

wisdom08 committed:
test: UserRepository 테스트 코드 작성 #73

brainbackdoor commented on src/test/java/com/insta/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,35 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
```
MySQL에 직접 테스트하나보군요~

wisdom08 replied:
일반적으로 h2 인메모리db 를 테스트용으로 사용하기는 하더라고요. 
다음에는 h2도 써봐야겠네요.

brainbackdoor commented on src/test/java/com/insta/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,35 @@
+@DataJpaTest
+class UserRepositoryTest {
+
+    public static final String USERNAME = "wisdom";
+    public static final String PASSWORD = "asdfasdf123";
+    public static final String EXCEPTION_MESSAGE_FOR_WRONG_USERNAME = "wrong username";
```
내부에서만 사용할 것이니 private으로 둬도 되지 않을까요

wisdom08 replied:
네 public 으로 둘 이유가 없네요. 확인 감사합니다.

brainbackdoor commented on src/test/java/com/insta/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,35 @@
+    public static final String PASSWORD = "asdfasdf123";
+    public static final String EXCEPTION_MESSAGE_FOR_WRONG_USERNAME = "wrong username";
+    @Autowired
+    private UserRepository userRepository;
```
Autowired를 사용한 필드주입과 생성자 주입의 차이에 대해 학습해보시겠어요?

wisdom08 replied:
면접 때 그 질문 여러 번 들었었는데 항상 생성자 주입이 권고하는 방식이고 권고하는 이유 몇 개 대답하긴 했었는데 좀 피상적으로만 알고 있는 느낌이에요.. 이번 주 주말에 말씀해주신 부분 정리 좀 해보겠습니다~!

wisdom08 replied:
[Autowired를 사용한 필드주입과 생성자 주입의 차이](https://github.com/wisdom08/my-wiki/blob/main/study/Autowired를%20사용한%20필드주입과%20생성자%20주입의%20차이.md)

brainbackdoor commented on src/test/java/com/insta/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,35 @@
+    void 저장된_유저_아이디로_조회() {
+        User user = User.createUser(USERNAME, PASSWORD);
+        userRepository.save(user);
+        User findUser = userRepository.findByUsername(USERNAME).orElseThrow(() -> new IllegalArgumentException(EXCEPTION_MESSAGE_FOR_WRONG_USERNAME));
```
- 테스트 코드에 로직을 작성하는건 지양해주세요.
- 위 테스트에서 Exception이 발생한다면, assertion으로 검증 가능한 상태인가요?
- 유저_저장 테스트와 중복 코드가 발생하는데요. 이를 제거해보는건 어떨까요

wisdom08 replied:
@brainbackdoor 
- 씨유님! 테스트코드에 로직을 작성하는 건 지양해달라고 하셨는데 어떤 로직을 말씀하시는 건가요? 
- 위 테스트에서는 익셉션이 발생해도 assertion 으로 검증이 가능한 상태가 아니네요... 제가 잘못 생각한 거 같아요. 현재 userService 에서 exist() 메서드로 유저가 있는지 체크하는 로직이 있어 userService 테스트를 작성할 때 해당 에러를 assertion 으로 검증해야할 것 같습니다. 
- 중복 코드 제거했습니다!

brainbackdoor replied:
- `.orElseThrow(() -> new IllegalArgumentException(EXCEPTION_MESSAGE_FOR_WRONG_USERNAME));` 이 부분이 로직이라고 생각헀어요~

wisdom08 replied:
아하.. 네 제거한 부분이네요!

brainbackdoor reviewed (approved):
피드백이 많이 늦었네요~ 😭
이슈 생성하고, 이를 기초로 작업 후 PR 주신 부분 좋았어요~
PR 메시지 만으로도 어떤 작업이 진행 중이고, 예정되어 있는지 알 수 있었어요~
몇가지 Comment 남겨두었는데 참고바래요~ 
오늘도 좋은 하루였길 바래요 🙏🏻

wisdom08 committed:
fix: private 접근 제한자로 변경함

- 현재 내부에서만 사용하기 때문

wisdom08 committed:
fix: `@BeforeEach` 활용해 중복코드 제거함

wisdom08 committed:
fix: 의존성 주입 방식을 필드주입에서 생성자주입으로 변경함

- immutable 객체로 사용, null 방지 등의 이유
- 참고: [Autowired 를 사용한 필드주입과 생성자 주입의 차이](https://github.com/wisdom08/my-wiki/blob/main/study/Autowired를%20사용한%20필드주입과%20생성자%20주입의%20차이.md)
